### PR TITLE
Add Lune library examples on endpoints

### PR DESCRIPTION
**Pull request by raguiar9080**
Showing examples of how one can use our TS/JS library to use the endpoints makes our library actually useful, since atm there is no direct connection between our documentation and our client.

For now, a snippet below the curl has been added to test and showcase, but we probably want to move this to a tabbed snippet that allows to either see the Curl example or the library one. The actual formatting of the snippet might also need some improvements.

The actual example is currently in vanilla JS, since trying to make it TS proved hard due to how we handle enums in our library. This might be an improvement to be made in the future.

Examples:
Screenshot 2023-03-21 at 16 16 20
Something that doesn't fit in snippet (have to scroll).
Screenshot 2023-03-21 at 16 17 17
Bunch of enums in request looks a bit ugly
Screenshot 2023-03-21 at 16 16 05
Blob example:
Screenshot 2023-03-21 at 16 19 54

### CODE REVIEW

**Comment by jstasiak** [comment]
Very nice work.

>     1. I went with a more JS approach on the example since it is the same form we use for the README in the `lune-ts` library.

I don't know how to interpret this sentence, what's "more JS" about this approach?

>     2. The initial `import` and setting up the `luneClient` seems a bit repetitive. However, it makes it possible to literally just copy paste things and it works. But copying multiple things will error since repeated imports and variable definitions.

I think it's the right approach. Being able to copy a snippet and run it = great success. People will (hopefully) understand that if they already have the setup code somewhere they only need to copy the actual method call 🤷‍♀️ 

>     3. Already present on the code as a TODO, but enum values will require a bit more work if we want to use them correctly. I'm simply converting to `any` atm 🙏 .

How about `"km" as "km"` or `"km" as const`? Both of these are just as easy as `"km" as any`.

>     4. For binary formats, I'm also simply converting to any. We might want to actually add something like readFile and send them, but I wanted to avoid that for now. I peeked at Stripe example and they do a readFile but it makes things a bit less copy pastable.

A `Blob` is expected there so we might just create a `Blob` instance? https://developer.mozilla.org/en-US/docs/Web/API/Blob

**Comment by raguiar9080** [comment]
> I don't know how to interpret this sentence, what's "more JS" about this approach?

I meant here that I only use Vanilla JS. No TS or other annotations. So everything ends with semicolon, use `var` etc.

> I think it's the right approach. Being able to copy a snippet and run it = great success. People will (hopefully) understand that if they already have the setup code somewhere they only need to copy the actual method call 🤷‍♀️

👍 

> How about `"km" as "km"` or `"km" as const`? Both of these are just as easy as `"km" as any`.

Unfortunately, none of these work according to my tests. I'm testing copying the snippets to dashboard repo btw. Since these values are enums, the types don't match.
I'm not sure if there's a way to do it without properly knowing the enum name itself. Again, this can be done correctly with the enum name itself, but will require some more significant changes since figuring out the enum requires us to keep track of previous steps while iterating to find the example.

However, correct me if I'm wrong but this is only a problem in TS, so I'm wondering if we should just remove the `as any` conversion. This would mean that copy and paste wouldn't work on TS though.

> A `Blob` is expected there so we might just create a `Blob` instance? https://developer.mozilla.org/en-US/docs/Web/API/Blob

I thought about that but usually Blobs are used in the context of a file or stream. Directly constructing a Blob is not that usual? But it is better than the current setup so I'll move to building a Blob  👍

**Comment by jstasiak** [comment]
@raguiar9080 

> Directly constructing a Blob is not that usual?

I mean when you interact with an API that expects it this is what you need to do I'd say

On the TS vs JS topic – I think we should encourage TS and discourage JS, but I'm not particularly attached to this. The only thing I think we shouldn't have is `as any` in the examples.

Do you have the compiler output from a case where you try `"km" as const`?

**Comment by raguiar9080** [comment]
> On the TS vs JS topic – I think we should encourage TS and discourage JS, but I'm not particularly attached to this. The only thing I think we shouldn't have is `as any` in the examples.

I agree that `as any` is quite ugly. But it will be a bit of work to make these use the actual correct enums so wondering if having it as future improvement is worth it.
Overall though, we seem to favour the JS way in our examples (that is the form we already showcase on README). I'm actually now thinking we should have both TS and JS examples being shown. 

> Do you have the compiler output from a case where you try `"km" as const`?

example:
```typescript
unit: "t" as const
```
gets me:
`[tsserver 2322] [E] Type '"t"' is not assignable to type 'MassUnit'.`

I'm not sure why tbh, I think it should work 🤷

**Comment by jstasiak** [comment]
Oh, *that* error. I actually like it. When I have an enum I want to use the enum to access the members, not hardcode the values somewhere.

Anyway, I think it's ok to have just JS examples for now

File: src/components/LuneTsExample.tsx [hunk 1]
```diff
@@ -0,0 +1,94 @@
+import Dereferencer from '@site/src/components/Dereferencer'
+import JsonPropertyParser from '@site/src/components/JsonPropertyParser'
+import ResourceExample from '@site/src/components/ResourceExample'
+import { AS_ANY_PLACEHOLDER, AS_BLOB_PLACEHOLDER, snakeToCamelCase } from '@site/src/utils'
+
+// Converts all properties that exist in the object to camelCase
+function camelCaseProperties(data: any): any {
```
Comment (jstasiak) [inline src/components/LuneTsExample.tsx:7]: See `camelCaseKey`, it even provides the right and precise types (if I remember correctly)

File: src/components/LuneTsExample.tsx [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        .map((param) => `'${Object.values(param)[0]}'`)
+        .join(', ')
+
+    if (requestBody) {
+        // If there are path/route parameters already, we need to handle that.
+        methodParameters = methodParameters ? methodParameters.concat(', ') : methodParameters
+
+        const dereferencedRequestBody = Dereferencer(requestBody)
+        const requestBodyExample = ResourceExample(requestBody, true, true)
+        // Every property on Lune-TS is camelCase so convert it
```
Comment (jstasiak) [inline src/components/LuneTsExample.tsx:69]: I don't think we ever call it Lune-TS

File: src/components/LuneTsExample.tsx [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    const cleanedMethodParameters = methodParameters
+        // Replace AS_ANY_PLACEHOLDER for actual 'as any'
+        .replace(new RegExp(`${AS_ANY_PLACEHOLDER}"`, 'g'), `" as any`)
+        // Replace AS_BLOB_PLACEHOLDER for actual Blob construction
+        .replace(new RegExp(`"(.+)${AS_BLOB_PLACEHOLDER}"`, 'g'), `new Blob(["$1"])`)
+
+    return `import * as lune from '@lune-climate/lune';
+var luneClient = new lune.LuneClient('${apiKey ?? '<API_KEY>'}');
```
Comment (jstasiak) [inline src/components/LuneTsExample.tsx:91]: `var`? Not `let` or `const`?

File: src/components/ResourceExample.tsx [hunk 4]
```diff
@@ -14,10 +15,32 @@ function isHighLevelElement(property: any): boolean {
 
 // TODO I believe the format can be used to provide better examples.
 // eslint-disable-next-line complexity
-function createRawExampleProperty(property: any, omitNotRequired: boolean): any {
+function createRawExampleProperty(
+    property: any,
+    omitNotRequired: boolean,
+    isLuneTSExample: boolean,
```
Comment (jstasiak) [inline src/components/ResourceExample.tsx:21]: If we follow the naming scheme we use shouldn't this be `Ts`, not `TS`?

**Review by jstasiak — APPROVED** [review]

**Comment by raguiar9080** [comment]
@rbruggem Whenever the snippet component has tabs let me know so this can be integrated there 👍

**Pull request merged**